lib/util/list_files.js

```diff
@@ -0,0 +1,145 @@
+});
+
+/**
+ * Return a list the files in the specified `repo` and its *open* submodules,
+ * relative to the root of the repository, or to the optionally specified
+ * `relativePath` if provided.  Note that the order of the list is
+ * not defined.  Note that if `relativePath` is provided, no files will be
+ * listed that are not within that subdirectory.  The behavior is undefined
+ * unless `relativePath` (if provided) is a valid relative path and has no
+ * trailing slash.
+ *
+ * @async
+ * @param {NodeGit.Repository} repo
+ * @param {String}             [relativePath]
+ * @return [String]
+ */
+exports.listFiles = co.wrap(function *(repo, relativePath) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    if (undefined !== relativePath && null !== relativePath) {
+        assert.isString(relativePath);
+        assert.notEqual(relativePath, "");
+        assert.notEqual(relativePath[relativePath.length - 1], "/");
+        assert(!path.isAbsolute(relativePath));
+    }
+    const subNames = new Set(yield SubmoduleUtil.getSubmoduleNames(repo));
+    const openSubs = yield SubmoduleUtil.listOpenSubmodules(repo);
+
+    const result = [];
+
+    let withSlash;  // the relative path with a suffixed '/'
+    if (relativePath) {
+        withSlash = relativePath + "/";
+    }
+
+    /**
+     * Return true if the specified `filePath` is contained in `relativePath`
+     * and false otherwise.
+     */
+    function inRelativePath(filePath) {
+        return filePath.startsWith(withSlash);
+    }
+
+    /**
+     * Add the specified `filePath` to the `result` array if not filtered.  If
+     * there is a relative path, remove the relative path from `filePath`
+     * before adding to `result`.
+     */
+    function addPath(filePath) {
+        if (relativePath) {
+            if (inRelativePath(filePath)) {
+                const relative = path.relative(relativePath, filePath);
+                result.push(relative);
+            }
+        }
+        else {
+            result.push(filePath);
+        }
+    }
+
+    // First look at all the files in the meta-repo, ignoring submodules
+    // `.gitmodules` file.
+
+    (yield listFilesForRepo(repo)).forEach(name => {
+        if (!subNames.has(name) &&
+            SubmoduleConfigUtil.modulesFileName !== name) {
+            addPath(name);
+        }
+    });
+
+    // Then list files in submodules.
+
+    yield openSubs.map(co.wrap(function *(name) {
+
+        // If we have a relative path, optimize by skipping submodules whose
+        // contents would be omitted.  Include the contents of this submodule
+        // if the relative path is one of the roots of the submodules path (its
+        // name starts with relative path), or if the relative path is actually
+        // inside the submodule (the submodule name is a prefix of the relative
+        // path).
+
+        if (relativePath &&
+           !(inRelativePath(name) || relativePath.startsWith(name))) {
```

Comment:
Indeed it does! I'll fix, but I don't think I can get a test case for it as it's just an optimization.